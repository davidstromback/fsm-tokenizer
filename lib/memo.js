import { createTokenizer } from "./tokenize.js";
import { text } from "./state.js";
import { defaultTokenFactory } from "./token.js";
const createMemoizedTokenizer = function tokenizeMemo(createToken = defaultTokenFactory) {
    const cache = {};
    return function* tokenizeMemo(input, initialState = text) {
        const key = `${initialState.name}:${input}`;
        const tokenize = createTokenizer(createToken);
        if (key in cache) {
            for (const token of cache[key].tokens) {
                yield token;
            }
            return cache[key].state;
        }
        const generator = tokenize(input, initialState);
        const tokens = [];
        let result = generator.next();
        while (!result.done) {
            tokens.push(result.value);
            yield result.value;
            result = generator.next();
        }
        const state = result.value;
        cache[key] = { state, tokens };
        return state;
    };
};
export { createMemoizedTokenizer };
