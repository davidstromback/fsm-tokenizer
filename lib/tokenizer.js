import { write } from "./context.js";
import { add, assign, clone, empty, point } from "./point.js";
const matchWhitespace = /\s/;
const matchLineBreak = /\n/;
function defaultTokenFactory(type, value, start, end, offset) {
    return {
        type,
        value,
        start: add(clone(start), offset),
        end: add(clone(end), offset),
    };
}
const defaultOptions = {};
export function tokenizer(schema, createToken = defaultTokenFactory) {
    return function* tokenize(input, options = defaultOptions) {
        var _a;
        const { state = schema.initialState, offset = empty } = options;
        const context = {
            char: "",
            contentEnd: point(-1),
            contentStart: point(),
            createToken,
            location: point(),
            next: undefined,
            offset,
            paddingEnd: point(-1),
            paddingStart: point(),
            state: schema.states[state],
            string: input,
            token: undefined,
        };
        for (let index = 0; index < input.length; index++) {
            context.char = input[index];
            context.token = undefined;
            if (context.next) {
                write(context);
                context.state = context.next;
                context.next = undefined;
            }
            context.state = context.state(context);
            assign(context.paddingEnd, context.location);
            context.location.offset++;
            if (matchLineBreak.test(context.char)) {
                context.location.column = 0;
                context.location.line++;
            }
            else {
                context.location.column++;
            }
            if (matchWhitespace.test(context.char)) {
                if (context.contentEnd.offset < context.paddingStart.offset) {
                    assign(context.contentStart, context.location);
                }
            }
            else {
                assign(context.contentEnd, context.paddingEnd);
            }
            if (context.token) {
                yield context.token;
            }
        }
        write(context);
        if (context.token) {
            yield context.token;
        }
        return {
            state: ((_a = context.next) !== null && _a !== void 0 ? _a : context.state).key,
            offset: add(context.location, context.offset),
        };
    };
}
