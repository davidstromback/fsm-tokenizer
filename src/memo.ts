import type { Tokenizer, TokenizerResult } from "./types.js";

import { empty } from "./point.js";

interface CacheEntry {
  result: TokenizerResult<string>;
  tokens: Array<unknown>;
}

const defaultOptions = {};

export function memo<StateKey extends keyof any, T>(
  tokenize: Tokenizer<T, StateKey>
): Tokenizer<T, StateKey>;

export function memo(
  tokenize: Tokenizer<unknown, string>
): Tokenizer<unknown, string> {
  const cache: Record<string, CacheEntry> = {};

  return function* tokenizeMemo(input: string, options = defaultOptions) {
    const { state = "", offset: { offset, line, column } = empty } = options;
    
    const key = `${offset}:${line}:${column}:${state}:${input}`;

    if (key in cache) {
      for (const token of cache[key].tokens) {
        yield token;
      }

      return cache[key].result;
    }

    const generator = tokenize(input, options);

    const tokens: Array<unknown> = [];

    let result = generator.next();
    while (!result.done) {
      tokens.push(result.value);
      yield result.value;
      result = generator.next();
    }

    cache[key] = { result: result.value, tokens };

    return result.value;
  };
}
