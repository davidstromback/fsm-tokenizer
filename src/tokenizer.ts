import type {
  Point,
  Token,
  TokenFactory,
  Context,
  Tokenizer,
  Schema,
} from "./types.js";

import { write } from "./context.js";
import { assign, clone, empty, point } from "./point.js";

const matchWhitespace = /\s/;
const matchLineBreak = /\n/;

function defaultTokenFactory<T>(
  type: T,
  value: string | undefined,
  start: Point,
  end: Point
): Token<T> {
  return { type, value, start: clone(start), end: clone(end) };
}

const defaultOptions = {};

export function tokenizer<
  TokenType extends keyof any,
  StateKey extends keyof any,
  T = Token<TokenType>
>(
  schema: Schema<TokenType, StateKey>,
  createToken?: TokenFactory<T, TokenType>
): Tokenizer<T, StateKey>;

export function tokenizer(
  schema: Schema<string, string>,
  createToken = defaultTokenFactory
): Tokenizer<unknown, string> {
  return function* tokenize(input, options = defaultOptions) {
    const { state = schema.initialState, offset = empty } = options;

    const context: Context = {
      char: "",
      contentEnd: point(-1),
      contentStart: clone(offset),
      createToken,
      location: clone(offset),
      next: undefined,
      offset: offset.offset,
      paddingEnd: point(-1),
      paddingStart: clone(offset),
      state: schema.states[state],
      string: input,
      token: undefined,
    };

    for (let index = 0; index < input.length; index++) {
      context.char = input[index];
      context.token = undefined;

      if (context.next) {
        write(context);
        context.state = context.next;
        context.next = undefined;
      }

      context.state = context.state(context);

      assign(context.paddingEnd, context.location);

      context.location.offset++;

      if (matchLineBreak.test(context.char)) {
        context.location.column = 0;
        context.location.line++;
      } else {
        context.location.column++;
      }

      if (matchWhitespace.test(context.char)) {
        if (context.contentEnd.offset < context.paddingStart.offset) {
          assign(context.contentStart, context.location);
        }
      } else {
        assign(context.contentEnd, context.paddingEnd);
      }

      if (context.token) {
        yield context.token;
      }
    }

    write(context);

    if (context.token) {
      yield context.token;
    }

    return {
      state: (context.next ?? context.state).key,
      offset: context.location,
    };
  };
}
