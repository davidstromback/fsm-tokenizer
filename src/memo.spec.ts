import { markup } from "./markup.js";
import { memo } from "./memo.js";
import { tokenizer } from "./tokenizer.js";
import { valid, invalid } from "./__tests__/cases.js";
import { tokenizeResult } from "./__tests__/fixtures.js";

describe("tokenize (memo)", () => {
  describe("returns the same tokens on subsequent calls", () => {
    test.each(valid)("%p", (input) => {
      const tokenizeMemo = memo(tokenizer(markup));
      const expected = tokenizeResult(tokenizeMemo, input);
      const actual = tokenizeResult(tokenizeMemo, input);

      expect(actual).toEqual(expected);

      actual.tokens.forEach((token, index) =>
        expect(token).toBe(expected.tokens[index])
      );
    });
  });

  describe("memoized results are equal to unmemoized results", () => {
    test.each(valid)("%p", (input) => {
      const tokenize = tokenizer(markup);
      const tokenizeMemo = memo(tokenizer(markup));

      expect(tokenizeResult(tokenizeMemo, input)).toEqual(
        tokenizeResult(tokenize, input)
      );

      expect(tokenizeResult(tokenizeMemo, input)).toEqual(
        tokenizeResult(tokenize, input)
      );
    });
  });

  describe("throws errors matching snapshots", () => {
    test.each(invalid)("%p", (input) => {
      const tokenizeMemo = memo(tokenizer(markup));

      expect(() => [...tokenizeMemo(input)]).toThrowErrorMatchingSnapshot();
    });
  });
});
