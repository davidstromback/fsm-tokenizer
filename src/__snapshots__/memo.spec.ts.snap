// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tokenize (memo) throws errors matching snapshots "<<" 1`] = `"Unexpected \\"<\\" in token \\"name\\", expected \\"\\\\s\\", \\"[a-zA-Z]\\", \\"\\\\/\\" or \\">\\"."`;

exports[`tokenize (memo) throws errors matching snapshots "<\\"" 1`] = `"Unexpected \\"\\"\\" in token \\"name\\", expected \\"\\\\s\\", \\"[a-zA-Z]\\", \\"\\\\/\\" or \\">\\"."`;

exports[`tokenize (memo) throws errors matching snapshots "<0>" 1`] = `"Unexpected \\"0\\" in token \\"name\\", expected \\"\\\\s\\", \\"[a-zA-Z]\\", \\"\\\\/\\" or \\">\\"."`;

exports[`tokenize (memo) throws errors matching snapshots "<foo 0foo />" 1`] = `"Unexpected \\"0\\" in token \\"name\\", expected \\"\\\\s\\", \\"[a-zA-Z]\\", \\"\\\\/\\" or \\">\\"."`;
